extractor/filesystem/os/opkg/opkg.go

```diff
@@ -0,0 +1,285 @@
+	IncludeNotInstalled bool
+}
+
+// DefaultConfig returns the default configuration for the OPKG extractor.
+func DefaultConfig() Config {
+	return Config{
+		MaxFileSizeBytes:    defaultMaxFileSizeBytes,
+		IncludeNotInstalled: defaultIncludeNotInstalled,
+	}
+}
+
+// Extractor extracts packages from OPKG files.
+type Extractor struct {
+	stats               stats.Collector
+	maxFileSizeBytes    int64
+	includeNotInstalled bool
+}
+
+// New returns a OPKG extractor.
+//
+// For most use cases, initialize with:
+// ```
+// e := New(DefaultConfig())
+// ```
+func New(cfg Config) *Extractor {
+	return &Extractor{
+		stats:               cfg.Stats,
+		maxFileSizeBytes:    cfg.MaxFileSizeBytes,
+		includeNotInstalled: cfg.IncludeNotInstalled,
+	}
+}
+
+// Config returns the configuration of the extractor.
+func (e Extractor) Config() Config {
+	return Config{
+		Stats:               e.stats,
+		MaxFileSizeBytes:    e.maxFileSizeBytes,
+		IncludeNotInstalled: e.includeNotInstalled,
+	}
+}
+
+// Name of the extractor.
+func (e Extractor) Name() string { return Name }
+
+// Version of the extractor.
+func (e Extractor) Version() int { return 0 }
+
+// Requirements of the extractor.
+func (e Extractor) Requirements() *plugin.Capabilities { return &plugin.Capabilities{} }
+
+// FileRequired checks if the specified file matches OPKG status file pattern.
+func (e Extractor) FileRequired(api filesystem.FileAPI) bool {
+	path := api.Path()
+	if !fileRequired(path) {
+		return false
+	}
+
+	fileinfo, err := api.Stat()
+	if err != nil {
+		return false
+	}
+	if e.maxFileSizeBytes > 0 && fileinfo.Size() > e.maxFileSizeBytes {
+		e.reportFileRequired(path, fileinfo.Size(), stats.FileRequiredResultSizeLimitExceeded)
+		return false
+	}
+
+	e.reportFileRequired(path, fileinfo.Size(), stats.FileRequiredResultOK)
+	return true
+}
+
+func fileRequired(path string) bool {
+	// status file
+	return filepath.ToSlash(path) == "usr/lib/opkg/status"
+}
+
+func (e Extractor) reportFileRequired(path string, fileSizeBytes int64, result stats.FileRequiredResult) {
+	if e.stats == nil {
+		return
+	}
+	e.stats.AfterFileRequired(e.Name(), &stats.FileRequiredStats{
+		Path:          path,
+		Result:        result,
+		FileSizeBytes: fileSizeBytes,
+	})
+}
+
+// Extract extracts packages from OPKG status files passed through the scan input.
+func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	inventory, err := e.extractFromInput(ctx, input)
+	if e.stats != nil {
+		var fileSizeBytes int64
+		if input.Info != nil {
+			fileSizeBytes = input.Info.Size()
+		}
+		e.stats.AfterFileExtracted(e.Name(), &stats.FileExtractedStats{
+			Path:          input.Path,
+			Result:        filesystem.ExtractorErrorToFileExtractedResult(err),
+			FileSizeBytes: fileSizeBytes,
+		})
+	}
+	return inventory, err
+}
+
+func (e Extractor) extractFromInput(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
```

Comment:
This function is almost identical to dpkg.go's extractFromInput except for the special cases for status.d.

How about we move the dpkg code to a shared library such as `extractor/filesystem/os/dpkg/parse/parse.go`, and call it from both dpkg.go and opkg.go? e.g. with a function signature like
`extractFromInput(ctx context.Context, input *filesystem.ScanInput, bool isDPKGStatusD) ([]*extractor.Inventory, error)`

Reply:
It's fine to modify the already in-place DPKG code, the unit tests should catch any major regressions. Since the package format is identical I expect future developments to be doable through the shared library as well and I'd rather not have a duplicate version of 100 lines of complex code in the codebase. So I'd say let's factor this out into shared code and if in the future they need to diverge we can split them then.

Reply:
Another option, similar to what [Syft](https://sourcegraph.com/github.com/anchore/syft/-/blob/syft/pkg/cataloger/debian/cataloger.go?L17) does, is to make the DPKG extractor be able to extract OPKGs as well - i.e. modify its FileRequired function and make sure we return the correct PURLs for OPKG packages, etc.